File: Difficulty: Medium/Directed Graph Cycle/directed-graph-cycle.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class DriverClass {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t-- > 0) {
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            int V = sc.nextInt();
            int E = sc.nextInt();
            for (int i = 0; i < V; i++)
                list.add(i, new ArrayList<Integer>());
            for (int i = 0; i < E; i++) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                list.get(u).add(v);
            }
            if (new Solution().isCyclic(V, list) == true)
                System.out.println("1");
            else
                System.out.println("0");
        }
    }
}
// } Driver Code Ends


/*Complete the function below*/

class Solution {
    public boolean isCyclic(int V, ArrayList<ArrayList<Integer>> adj) {
      boolean[] vis = new boolean[V];
      boolean[] pvis = new boolean[V];
     for(int i=0;i<V;i++) {
         if(!vis[i]) {
             
             if(dfs(i, -1, adj, vis, pvis)) {
                 return true;
             }
         }
     }
      return false;
    }
    public boolean dfs(int src, int par, ArrayList<ArrayList<Integer>> adj, 
    boolean[] vis, boolean pvis[]) {
        vis[src] = true;
        pvis[src] = true;
        for(int a: adj.get(src)) {
            if(!vis[a]) {
                if(dfs(a, src, adj, vis, pvis)) return true;
            } else if(pvis[a]) {
                return true;
            }
        }
        pvis[src] = false;
        return false;
    }
}

File: Difficulty: Medium/Directed Graph Cycle/directed-graph-cycle.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    
    bool dfs(int node, vector<bool>& visit, unordered_map<int, vector<int>>& adj, vector<bool>& st){
        if(st[node])
            return true;
        if(visit[node])
           return false;
           
        visit[node] = true;
        st[node] = true;
        
        for(auto neighbor : adj[node]){
            if(dfs(neighbor, visit, adj, st)){
                return true;
            }
        }
        st[node] = false;
        return false;
    }
  
    bool isCyclic(int V, vector<vector<int>> &edges) {
        // code here
        unordered_map<int, vector<int>> adj;
        for(auto& edge : edges){
            adj[edge[0]].push_back(edge[1]);
        }
        
        vector<bool> visit(V, false);
        vector<bool> st(V, false);
        
        for(int i=0; i<V; i++){
            if(!visit[i] && dfs(i, visit, adj, st)){
                return true;
            }
        }
        return false;
    }
};


//{ Driver Code Starts.

int main() {
    int tc;
    cin >> tc;
    cin.ignore();
    while (tc--) {
        int V, E;
        cin >> V >> E;
        cin.ignore();
        vector<vector<int>> edges;
        for (int i = 1; i <= E; i++) {
            int u, v;
            cin >> u >> v;
            edges.push_back({u, v});
        }

        Solution obj;
        bool ans = obj.isCyclic(V, edges);
        if (ans)
            cout << "true\n";
        else
            cout << "false\n";
    }
    return 0;
}
// } Driver Code Ends